data-model/src/main/java/org/hypertrace/core/datamodel/shared/StructuredTraceGraph.java

```diff
@@ -113,40 +118,48 @@ private void buildParentChildRelationship(StructuredTrace trace) {
       return;
     }
 
-    for (Entity entity : entities) {
-      entityMap.put(entity.getEntityId(), entity);
-    }
-
     List<Edge> entityEdges = trace.getEntityEdgeList();
     if (entityEdges != null) {
       for (Edge entityEdge : trace.getEntityEdgeList()) {
         Integer sourceIndex = entityEdge.getSrcIndex();
         Integer targetIndex = entityEdge.getTgtIndex();
         Entity parentEntity = entities.get(sourceIndex);
         Entity childEntity = entities.get(targetIndex);
-        parentToChildrenEntities.putIfAbsent(parentEntity, new ArrayList<>());
-        parentToChildrenEntities.get(parentEntity).add(childEntity);
-        childToParentEntities.putIfAbsent(childEntity, new ArrayList<>());
-        childToParentEntities.get(childEntity).add(parentEntity);
+        parentToChildrenEntities.putIfAbsent(parentEntity.getEntityId(), new ArrayList<>());
+        parentToChildrenEntities.get(parentEntity.getEntityId()).add(childEntity);
+        childToParentEntities.putIfAbsent(childEntity.getEntityId(), new ArrayList<>());
+        childToParentEntities.get(childEntity.getEntityId()).add(parentEntity);
```

Comment:
This can be combined to `childToParentEntities.computeIfAbsent(childEntity.getEntityId(), v -> new ArrayList<>()).add(parentEntity)`;

Reply:
Updated it with the suggested change.